homeassistant/components/rainforest_raven/config_flow.py

```
@@ -0,0 +1,159 @@
+        await self.async_set_unique_id(unique_id)
+        try:
+            await self._validate_device(dev_path)
+        except asyncio.TimeoutError:
+            return self.async_abort(reason="timeout_connect")
+        except (ParseError, SerialException):
+            return self.async_abort(reason="cannot_connect")
+        return await self.async_step_meters()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initiated by the user."""
+        if self._async_in_progress():
+            return self.async_abort(reason="already_in_progress")
+        ports = await self.hass.async_add_executor_job(serial.tools.list_ports.comports)
+        existing_devices = [
+            entry.data[CONF_DEVICE] for entry in self._async_current_entries()
+        ]
+        unused_ports = [
+            usb.human_readable_device_name(
+                port.device,
+                port.serial_number,
+                port.manufacturer,
+                port.description,
+                port.vid,
+                port.pid,
+            )
+            for port in ports
+            if port.device not in existing_devices
+        ]
+        if not unused_ports:
+            return self.async_abort(reason="no_devices_found")
+
+        errors = {}
+        if user_input is not None and user_input.get(CONF_DEVICE, "").strip():
+            port = ports[unused_ports.index(str(user_input[CONF_DEVICE]))]
+            dev_path = await self.hass.async_add_executor_job(
+                usb.get_serial_by_id, port.device
+            )
+            unique_id = _generate_unique_id(port)
+            await self.async_set_unique_id(unique_id)
+            try:
+                await self._validate_device(dev_path)
+            except asyncio.TimeoutError:
+                errors[CONF_DEVICE] = "timeout_connect"
+            except (ParseError, SerialException):
```

Exceptions like these should be handled in the third-party library and the code that raises them.

---

Can you clarify what you're looking to see here? The underlying library needs to express communication errors somehow. Would you like to see errors like this surface under a library-specific exception class?

---

The library should raise an specific error for these exceptions. These exceptions are implementation details and they should be transparent for the caller.

In our [developers docs](https://developers.home-assistant.io/docs/api_lib_index) we have written the following:

> One of the foundational rules of Home Assistant is that we do not include any protocol specific code. Instead, this code should be put into a standalone Python library and published to PyPI.

---

Okay, the library should only emit RAVEn-specific exceptions now.